### PR TITLE
add USART example

### PR DESCRIPTION
**Pull request by nebelgrau77**

### CODE REVIEW

File: examples/serial.rs [hunk 1]
```diff
@@ -0,0 +1,54 @@
+#![no_main]
+#![no_std]
+
+use panic_halt;
```
Comment (therealprof) [inline examples/serial.rs:4]:
```suggestion
use panic_halt as _;
```

File: examples/serial.rs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    prelude::*,
+    stm32,
+    serial::config::Config,
+    serial::Serial};
+
+use core::fmt::Write; // for pretty formatting of the serial output
+
+#[entry]
+fn main() -> ! {
+
```
Comment (therealprof) [inline examples/serial.rs:20]:
```suggestion
```

**Review by therealprof — COMMENTED** [review]
If I'm not mistaken this HAL impl should allow you to `writeln!` directly onto the `serial`. Splitting should only be necessary if you want to read and write in different parts of your application.

**Comment by birkenfeld** [comment]
Why not show receiving data as well? That's usually the more tricky part...

**Comment by therealprof** [comment]
@birkenfeld The more examples the merrier... ;) I don't see any problem with a simple example demonstrating a very common usecase (like printing logging info) and a more complex one doing more fancy stuff like interacting with an external device.

**Review by nebelgrau77 — COMMENTED** [review]
Modified as suggested.

**Comment by nebelgrau77** [comment]
> If I'm not mistaken this HAL impl should allow you to `writeln!` directly onto the `serial`. Splitting should only be necessary if you want to read and write in different parts of your application.

I guess my n00bness strikes again: I saw that in other examples and assumed you actually had to split. But then again, commenting the `split` line and replacing `tx` with `serial` (I guess you meant that) will not compile: 

```rust
writeln!(serial, "value: {}\r", value).unwrap();
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ method not found in `stm32f4xx_hal::serial::Serial<stm32f4xx_hal::stm32::USART2, (stm32f4xx_hal::gpio::gpioa::PA2<stm32f4xx_hal::gpio::Alternate<stm32f4xx_hal::gpio::AF7>>, stm32f4xx_hal::gpio::gpioa::PA3<stm32f4xx_hal::gpio::Alternate<stm32f4xx_hal::gpio::AF7>>)>`
```

**Comment by therealprof** [comment]
> > If I'm not mistaken this HAL impl should allow you to `writeln!` directly onto the `serial`. Splitting should only be necessary if you want to read and write in different parts of your application.
> 
> I guess my n00bness strikes again: I saw that in other examples and assumed you actually had to split. But then again, commenting the `split` line and replacing `tx` with `serial` (I guess you meant that) will not compile:
> 
> ```rust
> writeln!(serial, "value: {}\r", value).unwrap();
>    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ method not found in `stm32f4xx_hal::serial::Serial<stm32f4xx_hal::stm32::USART2, (stm32f4xx_hal::gpio::gpioa::PA2<stm32f4xx_hal::gpio::Alternate<stm32f4xx_hal::gpio::AF7>>, stm32f4xx_hal::gpio::gpioa::PA3<stm32f4xx_hal::gpio::Alternate<stm32f4xx_hal::gpio::AF7>>)>`
> ```

Weird. No, not your fault, the implementation is missing. :(

**Comment by nebelgrau77** [comment]
> Why not show receiving data as well? That's usually the more tricky part...

Mainly because I don't know, how :) I just wanted to add a simple example that I use a lot, and that was missing. For other beginners :)

**Comment by therealprof** [comment]
If you could run `cargo fmt` and squash the commits, we can merge this.

**Comment by therealprof** [comment]
> Is it to satisfy clippy, because otherwise it will suggest to remove that entirely?

Yes, the compiler (and hence clippy) cannot see that anything from that crate is actually used (because it contains a panic handler which will only be added very late in the compilation process) so it is important to `use` it but by renaming it to `_` the compiler (and clippy) won't whine about it.

**Comment by nebelgrau77** [comment]
> If you could run `cargo fmt` and squash the commits, we can merge this.

I'd love to but I'm not sure, how: I see "Merging is blocked: Merging can be performed with 1 approving review" - unless it has nothing to do with that?

**Comment by therealprof** [comment]
In a nutshell I'd ask you to:
- Run `cargo fmt`
- Commit the resulting change
- Do a `git reset --soft origin/master`
- Do a `git commit -a`
- Do a git push --force onto your PR branch

**Comment by therealprof** [comment]
@nebelgrau77 Close. 😅 The history was messed up so to be on the safe side I rebased your branch.

**Review by therealprof — APPROVED** [review]

**Comment by therealprof** [comment]
Thanks for sticking with us.

**Comment by nebelgrau77** [comment]
> @nebelgrau77 Close.  The history was messed up so to be on the safe side I rebased your branch.

Thanks! I knew it was to good to be true :D

**Pull request merged**